src/include/transaction/deferred_action_manager.h

```diff
@@ -0,0 +1,108 @@
+#pragma once
+#include <queue>
+#include <utility>
+#include <vector>
+#include "transaction/timestamp_manager.h"
+#include "transaction/transaction_defs.h"
+
+namespace terrier::transaction {
+/**
+ * The deferred action manager tracks deferred actions and provides a function to process them
+ */
+class DeferredActionManager {
+ public:
+  /**
+   * Constructs a new DeferredActionManager
+   * @param timestamp_manager source of timestamps in the system
+   */
+  explicit DeferredActionManager(TimestampManager *timestamp_manager) : timestamp_manager_(timestamp_manager) {}
+
+  /**
+   * Adds the action to a buffered list of deferred actions.  This action will
+   * be triggered no sooner than when the epoch (timestamp of oldest running
+   * transaction) is more recent than the time this function was called.
+   * @param a functional implementation of the action that is deferred. @see DeferredAction
+   */
+  timestamp_t RegisterDeferredAction(const DeferredAction &a) {
+    timestamp_t result = timestamp_manager_->CurrentTime();
+    {
+      common::SpinLatch::ScopedSpinLatch guard(&deferred_actions_latch_);
+      new_deferred_actions_.emplace(result, a);
+    }
+    return result;
+  }
+
+  /**
+   * Adds the action to a buffered list of deferred actions.  This action will
+   * be triggered no sooner than when the epoch (timestamp of oldest running
+   * transaction) is more recent than the time this function was called.
+   * @param a functional implementation of the action that is deferred
+   */
+  timestamp_t RegisterDeferredAction(const std::function<void()> &a) {
+    // TODO(Tianyu): Will this be a performance problem? Hopefully C++ is smart enough
+    // to optimize out this call...
+    return RegisterDeferredAction([=](timestamp_t /*unused*/) { a(); });
+  }
+
+  /**
+   * Clear the queue and apply as many actions as possible
+   * @return numbers of deferred actions processed
+   */
+  uint32_t Process() {
+    uint32_t processed = 0;
+    // Check out a timestamp from the transaction manager to determine the progress of
+    // running transactions in the system.
+    timestamp_t oldest_txn = timestamp_manager_->OldestTransactionStartTime();
+    processed += ClearBacklog(oldest_txn);
+    processed += ProcessNewActions(oldest_txn);
```

Comment:
Because the queue may not be in time stamp order (see above), this call can result in out-of-order processing of some parts of the queue.  Shouldn’t processing new actions be conditional on backlog being empty?

Reply:
Why? I don't think we should guarantee in-order processing for deferred actions. If there are dependencies, users should probably not queue them in parallel. (i.e. they should defer something that defers the second thing.)

The code should still work if we wrap the following line in a condition. Just want to discuss why you think it's necessary for correctness.

Reply:
I’m not sure if it causes a correctness problem in practice, but our reasoning about double deferrals for drop index and drop table relies in part on the assumption that the deferral queue is processed in order.

We’re certainly making it significantly harder to reason about the deferrals, but I think in practice it only results in reordered execution of parallelizable tasks.  However, I cannot prove that at the moment and it would definitely require some strong analysis showing correctness for the paper.

Reply:
Leave it for now and open an issue?

Reply:
Sounds good.